### PR TITLE
kvserver: don't check ConnHealth when releasing proposal quota

### LINKED ISSUE #84202: kvserver: quota pool does not consider node liveness
The replica quota pool does not take node liveness into account. Instead, it only considers the RPC status:

https://github.com/cockroachdb/cockroach/blob/d35cf750ba6e11176a14cfb32e78049b77e84209/pkg/kv/kvserver/replica_proposal_quota.go#L167-L170

We should check the liveness record instead.

### PR DESCRIPTION
**Pull request by pav-kv**
The proposal quota release procedure checked node connection health for every
node that appears active after replica activity checks. This is expensive, and
previously caused issues like https://github.com/cockroachdb/cockroach/issues/84943.

This change removes the ConnHealth check, because other checks, such as
isFollowerActiveSince and paused replicas, provide sufficient protection from
various kinds of overloads.

Touches #84202

Release note: None

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
I'll work on a test tomorrow, but for now: is the approach correct?

File: pkg/kv/kvserver/replica_proposal_quota.go [hunk 1]
```diff
@@ -164,9 +164,12 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
...
 		}
-
-		// Only consider followers that that have "healthy" RPC connections.
-		if err := r.store.cfg.NodeDialer.ConnHealth(rep.NodeID, r.connectionClass.get()); err != nil {
+		// At this point, we know that either we communicated with this replica
+		// recently, or we became the leader recently. The latter case is ambiguous
+		// w.r.t. the actual state of that replica, but it will be disambiguated
+		// soon. We do a second check using the NodeLiveness to increase the chance
+		// of disambiguating here.
+		if !r.store.cfg.NodeLiveness.IsAvailable(rep.NodeID) {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_proposal_quota.go:172]: Alternatives:

1. Only do this check if we recently became the leader, because otherwise `isFollowerActiveSince` does the job.
1. Only do the `NodeLiveness` check. Are there cases when it would not detect things that per-replica activity detects?
1. Leave the `ConnHealth`, but gate it by "I recently became the leader", similarly to (1). Are there cases when `ConnHealth` is more accurate than `NodeLiveness`?

Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_proposal_quota.go:172]: `NodeLiveness` field is nil in some of the tests. Is it expected?

Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_proposal_quota.go:172]: I think the case we're interested in here is where the follower replica is active, and the node is up and running, but the node is struggling and is failing its liveness heartbeats. This is unrelated to whether we just became the leader or not -- even if we've been the leader for a long time, and `isFollowerActiveSince` returns `true`, if the node is struggling then we don't want it to hold up the range.

One typical example of this is a stalled disk, since we do a synchronous disk write during node heartbeats specifically to check this (stalled disks are a common failure mode with networked disks used in cloud VMs):

https://github.com/cockroachdb/cockroach/blob/eeb72360fe8e5bdc9ed123612c4e71b975b52124/pkg/kv/kvserver/liveness/liveness.go#L1286-L1288

So we should have both checks:

* `isFollowerActiveSince` will check that the Raft replica is live.
* `IsAvailable` will check that the node is live and reasonably healthy.

This is related to a broader discussion on whether we even want the quota pool to ever hold up the range at all (#75978). Most of the time, we really don't. But for now, this is one step in the right direction.

I'll also note that `IsAvailable()` will return false if the node is not in the local liveness cache (i.e. we don't know about it). This seems fine, since that will default to not holding up the range.

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_proposal_quota.go:172]: Do you know the condition under which it's nil? Is it only in tests, or can happen in the real binary too?

Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_proposal_quota.go:172]: I mean, it can be nil very early during server startup, but I'm pretty sure that we set it up before spinning up the Raft replicas.

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_proposal_quota.go:172]: Is this disk write on the requested node, or on some other range/nodes where the liveness records are stored?

Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_proposal_quota.go:172]: This happens on every node every time they send a heartbeat.

For reference, liveness records themselves are stored as key/value pairs in a range, just like any other data in the system. So there will be one leaseholder somewhere in the cluster who's responsible for them.

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_proposal_quota.go:172]: Ah, got it.

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_proposal_quota.go:172]: What kinds of events refresh `isFollowerActiveSince` timestamps? Are these successful Raft RPCs, or any interactions?
Any example when `isFollowerActiveSince` is true, but the node is not live?

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_proposal_quota.go]: Done.

Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_proposal_quota.go]: > What kinds of events refresh `isFollowerActiveSince` timestamps? Are these successful Raft RPCs, or any interactions?

Any Raft message received from the follower:

https://github.com/cockroachdb/cockroach/blob/d10bde7a099716fcc4c3270c67996899fb121a01/pkg/kv/kvserver/replica_raft.go#L517-L518

>Any example when `isFollowerActiveSince` is true, but the node is not live?

A simple but contrived example might be a node with a partial network partition: it's not able to communicate with the liveness leaseholder (so it can't heartbeat), but it is able to communicate with other nodes. This isn't all that relevant, because in this case the node itself isn't necessarily struggling as such, but it's a fairly clear example.

A more relevant example might be a very overloaded node. Maybe it's taking more that 4.5 seconds (the heartbeat timeout) to fsync the disk, so node heartbeats aren't being sent. Or maybe the heartbeat goroutine isn't getting scheduled or the network is backlogged. However, it is able to occasionally/eventually append Raft proposals and respond to Raft leaders. In these cases, replicas on that node would be very slow to append Raft proposals, and we would rather sacrifice the followers on that node than slow down the entire range.

**Comment by pavelkalinnikov** [comment]
How about an alternative simple approach:
- Collect states of the replicas (the same way as we do now), particularly the log size estimate from each.
- For those replicas that we didn't hear from, be pessimistic.
- Set the `minIndex` to be the median of the log size estimates (all of them, not just those that are alive; but the dead/slow ones will be in the bottom of the sorted list).

If a majority of replicas is healthy, we will progress with the median speed of them. This has less chance of being stuck, even if the "slowest" node is in fact inactive or fails to update liveness. With such an approach, we can avoid both `isFollowerActiveSince` and `NodeLiveness` check.

I guess, the problem with this approach is that the slowest replicas will increasingly fall behind. Are there any other throttling mechanisms that try slow down the quick replicas, so that all replicas can walk at the same speed?

**Comment by erikgrinaker** [comment]
What you're saying makes a lot of sense, and I think this is the direction we want to move in. Rather than checking the median and the log index, we'd maybe want to check that a quorum have applied up to some index. Raft already requires a quorum to have appended an entry before it is committed, which provides a baseline of throttling here.

We did something similar recently in #83851, where we stop replicating to followers that are experiencing IO overload (specifically, where Pebble compactions aren't able to keep up with incoming writes, inverting the LSM). But we make sure we keep a quorum of followers to make forward progress.

This is related to [admission control](https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/admission_control.md), which monitors the local Pebble store's health and begins backpressuring and prioritizing incoming traffic once Pebble starts struggling. Raft traffic is a bit trickier, in that the Raft log must be deterministic, so once a command has been committed to it then we can't drop or reorder commands. On the leaseholder, we can (and do) apply admission control to incoming traffic before it hits the Raft log, but on followers there's really nothing we can do -- we'll just have to append and apply those entries as-is.

So longer-term, the leaseholder needs to take the health of followers into account for purposes of admission control, and if followers are getting overloaded, we'll have to rely on the allocator to start moving them around (as you allude to in your comment). The role of the quota pool in all of this is still a bit unclear. There's been a lot of discussions about this recently, e.g. on #75066 and #83851. There are also related challenges with flow control of Raft messages, since we can currently OOM if the Raft queues fill up with huge messages (e.g. SSTable ingestion), see #79755.

For now though, we'll have to get rid of the useless RPC check anyway, and we may as well add in a liveness check while we're here. We're planning on doing more work around admission control and the quota pool for 23.1, but we haven't really firmed up a concrete plan yet (Tobi and Sumeer have been more involved in this than I have).

File: pkg/kv/kvserver/replica_proposal_quota.go [hunk 2]
```diff
@@ -164,9 +164,14 @@ func (r *Replica) updateProposalQuotaRaftMuLocked(
...
-
-		// Only consider followers that that have "healthy" RPC connections.
-		if err := r.store.cfg.NodeDialer.ConnHealth(rep.NodeID, r.connectionClass.get()); err != nil {
+		// At this point, we know that either we communicated with this replica
+		// recently, or we became the leader recently. The latter case is ambiguous
+		// w.r.t. the actual state of that replica, but it is temporary.
+		//
+		// We do a second check using the NodeLiveness to increase the chance
+		// of disambiguating here, and also to catch other slowness that may be
+		// surfaced in the liveness but not in isFollowerActiveSince.
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_proposal_quota.go:173]: I think we can reword this to allude to the fact that we don't want a node that isn't heartbeating (for whatever reason) to hold up the range.

**Review by erikgrinaker — COMMENTED** [review]
Did you and @aliher1911 come up with an easy way to test this? If not, I think this is fine, and we can rely on broader integration/end-to-end tests to pick up any issues.

PS: let's tack on " in quota pool" in the commit title.

File: pkg/kv/kvserver/replica_proposal_quota.go (hunk 2 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_proposal_quota.go]: Not needed, as we decided on not adding the `NodeLiveness` check.

**Review by erikgrinaker — APPROVED** [review]
LGTM, congrats on your first PR! 🚀

Let's change "Fixes #84202" to "Touches #84202", and add a comment on the issue explaining why we didn't do it this time around. It's something we may want to consider doing later, so let's leave it open. Also, the "Fixes/touches #84202" part isn't usually included in the commit message, only the PR description.

**Comment by pavelkalinnikov** [comment]
Note: This PR changed its focus from changing a check to just removing it. As a result, there are strictly fewer cases when the quota is hold up, so this change won't introduce new quota-related blockages. The risk is in introducing more cases when followers are overloaded, but there is a feedback mechanism ("paused replicas") which pushes back, so there is already a good protection from this.

Not adding tests.

**Comment by pavelkalinnikov** [comment]
> Let's change "Fixes #84202" to "Touches #84202", and add a comment on the issue explaining why we didn't do it this time around. It's something we may want to consider doing later, so let's leave it open. Also, the "Fixes/touches #84202" part isn't usually included in the commit message, only the PR description.

All done.

**Comment by erikgrinaker** [comment]
Let's backport this change too. You can add on the `backport-21.2.x` and `backport-22.1.x` labels, and Blathers will open the backport PRs for you when this merges (if the patch applies cleanly).

**Comment by pavelkalinnikov** [comment]
bors r=erikgrinaker
